agent/src/main.rs

```diff
@@ -92,20 +88,37 @@ async fn main() -> Result<()> {
         .layer(DefaultBodyLimit::max(MAX_BODY_SIZE_BYTES))
         .with_state(state);
 
-    let addr = format!("{}:{}", host, port);
-    log::info!("serval agent daemon listening on {}", &addr);
+    let predefined_port: Option<u16> = match std::env::var("PORT") {
+        Ok(port_str) => port_str.parse::<u16>().ok(),
+        Err(_) => None,
+    };
+
+    // Start the Axum server; this is in a loop so we can try binding more than once in case our
+    // randomly-selected port number ends up conflicting with something else due to a race condition.
+    let mut port: u16;
+    let server: Server<_, _> = loop {
+        port = predefined_port.unwrap_or_else(|| find_nearest_port(8100).unwrap());
+        let addr: SocketAddr = format!("{host}:{port}").parse().unwrap();
+        match axum::Server::try_bind(&addr) {
+            Ok(builder) => break builder.serve(app.into_make_service()),
```

Comment:
I don't love how easy it is to lose the creation of the server, which is in here. It is returned as the result of the loop by using the `break` keyword (which, I must say, is a really nifty Rust feature).